tests/Headers.hs

```diff
@@ -0,0 +1,90 @@
+{-# OPTIONS_GHC -fno-warn-type-defaults #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Headers where
+
+import qualified Data.ByteString.Char8 as BC
+import Data.Attoparsec.ByteString.Char8 (parseOnly)
+
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit ((@=?), testCase, Assertion)
+
+import Data.MIME
+
+
+unittests :: TestTree
+unittests =
+    testGroup
+        "address parsing tests"
+        [ parsesMailboxesSuccessfully
+        , parsesAddressesSuccessfully
+        ]
+
+-- | Note some examples are taken from https://tools.ietf.org/html/rfc3696#section-3
+mailboxes :: [(String, Either String Mailbox -> Assertion, BC.ByteString)]
+mailboxes =
+    [ ( "address with FQDN"
+      , (Right (Mailbox Nothing "foo@bar.com") @=?)
+      , BC.pack "<foo@bar.com>")
+    , ( "just with a host name"
+      , (Right (Mailbox Nothing "foo@bar") @=?)
+      , BC.pack "<foo@bar>")
+    , ( "domain as IPv4"
+      , (Right (Mailbox (Just "roman") "roman@192.168.1.1") @=?)
+      , BC.pack "roman <roman@[192.168.1.1]>")
+    , ( "domain as IPv6"
+      , (Right (Mailbox (Just "roman") "roman@::1") @=?)
```

Comment:
Just to reiterate, we have discarded any *explicit* indication that this was a domain literal.  It can most likely
be deduced, but formally, RFC 5322 has no knowledge of the address format - it can be any printable ascii except `[`, `]` and `\`.

I think it will be easier/safer to work with a tagged union anyway.